tools/Dashboard/DevHome.Dashboard/ComSafeWidgetObjects/ComSafeWidget.cs

```diff
@@ -0,0 +1,214 @@
+                return await OopWidget.GetCustomStateAsync();
+            }
+        });
+    }
+
+    public async Task<WidgetSize> GetSizeAsync()
+    {
+        return await Task.Run(async () =>
+        {
+            try
+            {
+                return await OopWidget.GetSizeAsync();
+            }
+            catch (COMException e) when (e.HResult == RpcServerUnavailable || e.HResult == RpcCallFailed)
+            {
+                await GetNewOopWidgetAsync();
+                return await OopWidget.GetSizeAsync();
+            }
+        });
+    }
+
+    public async Task NotifyActionInvokedAsync(string verb, string data)
+    {
+        await Task.Run(async () =>
+        {
+            try
+            {
+                await OopWidget.NotifyActionInvokedAsync(verb, data);
+            }
+            catch (COMException e) when (e.HResult == RpcServerUnavailable || e.HResult == RpcCallFailed)
+            {
+                await GetNewOopWidgetAsync();
+                await OopWidget.NotifyActionInvokedAsync(verb, data);
+            }
+        });
+    }
+
+    public async Task DeleteAsync()
+    {
+        await Task.Run(async () =>
+        {
+            try
+            {
+                await OopWidget.DeleteAsync();
+            }
+            catch (COMException e) when (e.HResult == RpcServerUnavailable || e.HResult == RpcCallFailed)
+            {
+                await GetNewOopWidgetAsync();
+                await OopWidget.DeleteAsync();
+            }
+        });
+    }
+
+    public async Task SetCustomStateAsync(string state)
+    {
+        await Task.Run(async () =>
+        {
+            try
+            {
+                await OopWidget.SetCustomStateAsync(state);
+            }
+            catch (COMException e) when (e.HResult == RpcServerUnavailable || e.HResult == RpcCallFailed)
+            {
+                await GetNewOopWidgetAsync();
+                await OopWidget.SetCustomStateAsync(state);
+            }
+        });
+    }
+
+    public async Task SetSizeAsync(WidgetSize widgetSize)
+    {
+        await Task.Run(async () =>
+        {
+            try
+            {
+                await OopWidget.SetSizeAsync(widgetSize);
+            }
+            catch (COMException e) when (e.HResult == RpcServerUnavailable || e.HResult == RpcCallFailed)
+            {
+                await GetNewOopWidgetAsync();
+                await OopWidget.SetSizeAsync(widgetSize);
+            }
+        });
+    }
+
+    public async Task NotifyCustomizationRequestedAsync()
+    {
+        await Task.Run(async () =>
+        {
+            try
+            {
+                await OopWidget.NotifyCustomizationRequestedAsync();
+            }
+            catch (COMException e) when (e.HResult == RpcServerUnavailable || e.HResult == RpcCallFailed)
+            {
+                await GetNewOopWidgetAsync();
+                await OopWidget.NotifyCustomizationRequestedAsync();

```

Comment:
Is there a risk of infinite loop here if it keeps failing?

Reply:
There's no loop -- the calls inside the method are both on the "real", OOP object. It's just try once, if it doesn't work try again, the end.

Reply:
oh perfect!